File: lib/jack.rb
require 'logger'
require 'rubygems'
require 'johnson'
require 'rack'
gem 'builder'

module Jack
  VERSION = '0.2.0'

  def self.run(app_file, options)
    @js = Johnson::Runtime.new
    @js.load(File.dirname(__FILE__) + "/jack.js")
    Dir.glob(File.dirname(__FILE__) + "/jack/*.js").each { |f| @js.load(f) }

    @log = Logger.new(options.log)
    @log.info "Starting #{app_file} on http://#{options.host}:#{options.port}"
    @js['Jack']['root'] = File.dirname(app_file)
    @js['Jack']['log']  = @log # Why can't we go the other way? (Jack.log = Ruby.Jack.log)
    @js['Jack']['options'] = {:Port => options.port, :Host => options.host}

    @js.load(app_file)
  end

  def self.log; @log end
end

File: lib/jack.js
Johnson.require("johnson/browser");
Johnson.require("johnson/browser/jquery");

var Jack = {
  version: Ruby.Jack.VERSION,
  default_headers: {'Content-Type': 'text/html'},
  
  up: function(){
    Ruby.Rack.Handler.Mongrel.run(
      function(env) {
          Ruby.Dir.glob("controllers/*.js").each(function(file) {
              eval(Ruby.File.read(file));
          });
        
          try {
              return Jack.Action.route(
                  env,
                  Jack.Utils.parameterize(env['QUERY_STRING'])
              );
          }
          catch(e) {
              return Jack.application_error(e);
          }
      },
      Jack.options || {}
    )
  },
  
  not_found: function () {
    return [404, Jack.default_headers, "<h1>Not Found</h1>"];
  },

  application_error: function (e) {
    var info = e.inspect ? e.inspect() : e.name + ": " + e.message
    Jack.log.error(info);
    var html = ["<h1>Application Error: " + info + "</h1>",
      "<p>" + e.toString() + "</p>"].join("\n");
    return [500, Jack.default_headers, html];
  }
};
